crates/sui-archival/src/reader.rs

```diff
@@ -257,6 +258,115 @@ impl ArchiveReader {
+                                Ok::<(), anyhow::Error>(())
+                            })
+                    });
+                futures::future::ready(result)
+            })
+            .await
+    }
+
+    /// Load checkpoints+txns+effects from archive into the input store `S` for the given
+    /// checkpoint range. If latest available checkpoint in archive is older than the start of the
+    /// input range then this call fails with an error otherwise we load as many checkpoints as
+    /// possible until the end of the provided checkpoint range.
+    pub async fn read_summaries<S>(
+        &self,
+        store: S,
+        checkpoint_range: Range<CheckpointSequenceNumber>,
+        checkpoint_counter: Arc<AtomicU64>,
+    ) -> Result<()>
+    where
+        S: WriteStore + Clone,
+        <S as ReadStore>::Error: std::error::Error,
+    {
+        let (summary_files, start_index, end_index) =
+            self.get_summary_files(checkpoint_range.clone()).await?;
+        let remote_object_store = self.remote_object_store.clone();
+        futures::stream::iter(summary_files.iter())
+            .enumerate()
+            .filter(|(index, _s)| future::ready(*index >= start_index && *index < end_index))
+            .map(|(_, summary_metadata)| {
+                let remote_object_store = remote_object_store.clone();
+                async move {
+                    let summary_data =
+                        get(&summary_metadata.file_path(), remote_object_store.clone()).await?;
+                    Ok::<Bytes, anyhow::Error>(summary_data)
+                }
+            })
+            .boxed()
+            .buffered(self.concurrency)
+            .try_for_each(|summary_data| {
+                let result: Result<(), anyhow::Error> =
+                    make_iterator::<CertifiedCheckpointSummary, Reader<Bytes>>(
+                        SUMMARY_FILE_MAGIC,
+                        summary_data.reader(),
+                    )
+                    .and_then(|summary_iter| {
+                        summary_iter
+                            .filter(|s| {
+                                s.sequence_number >= checkpoint_range.start
+                                    && s.sequence_number < checkpoint_range.end
+                            })
+                            .try_for_each(|summary| {
+                                let verified_checkpoint =
+                                    Self::get_or_insert_verified_checkpoint(&store, summary)?;
+                                // Update highest synced watermark
+                                store
+                                    .update_highest_verified_checkpoint(&verified_checkpoint)
+                                    .map_err(|e| anyhow!("Failed to update watermark: {e}"))?;
+                                checkpoint_counter.fetch_add(1, Ordering::Relaxed);
+                                Ok::<(), anyhow::Error>(())
```

Comment:
This seems to be the main difference between this fn and the unverified. Could we refactor the common bits? I could be missing something though.

Reply:
The other major difference is also calling `buffered` vs `buffered_unordered`. I'll try to see if i can refactor those two things, but it might be tricky

Reply:
Oh I see. Feel free to leave as is if it ends up being tricky